### PR TITLE
Fix revalidation storage (#239)

### LINKED ISSUE #239: Potential bug : Cache is not updated when a revalidation request is issued (AsyncTransport, but synchronous CacheTransport looks the same)
Hello,

First, thanks for the excellent work on hishel, the codebase is a pleasure to work with.

I'm running into what I believe is an issue (but may very well be a misunderstanding from me).

My setup uses an asynchronous httpx client, and thus I'm also using the async hishel interface (although it looks that the synchronous interface use the same algorithm). `hishel.Controller` is configured with `allow_heuristics=True` and `allow_stale=True` (although I'm not sure that matters, as what I see seems to come from the transport implementation).

I'm making simple requests to an httpbin's "cache/10" endpoint, which is a simple endpoint that returns a response containing a `cache-control: public, max-age=10` header.

The first request is made and the response is returned as-is by hishel (as expected), and the cache entry is created (as expected).
The second request (before max-age expires) is returned from cache, all good too.
But now if I wait for the max-age to expire, the next request will get revalidated (which is expected, I guess, or at least that looks logical to me). But this revalidation request, that gets once again the same response (max-age 10), is not stored into cache.
All future requests will be revalidated too, making the cache for this cache key completely ignored forever from this point on.

One graphical representation of the sequence is the following (the little icon next to the request time shows whether or not the response has been sent from hishel cache, time goes from bottom to top:

image

I was able to "fix" the behaviour that I think is not what we expect by adding the following fast and dirty patch to the async transport:

```
diff --git a/hishel/_async/_transports.py b/hishel/_async/_transports.py
index 5d01274..82b9183 100644
--- a/hishel/_async/_transports.py
+++ b/hishel/_async/_transports.py
@@ -186,6 +186,15 @@ class AsyncCacheTransport(httpx.AsyncBaseTransport):
                 await revalidation_response.aclose()

                 assert isinstance(final_httpcore_response.stream, tp.AsyncIterable)
+
+                if self._controller.is_cachable(
+                    request=httpcore_request, response=final_httpcore_response
+                ):
+                    await self._storage.store(
+                        key,
+                        response=final_httpcore_response,
+                        request=httpcore_request,
+                    )
                 return await self._create_hishel_response(
                     key=key,
                     response=final_httpcore_response,
```

With the patch applied, I get the behaviour I'm expecting which is that one request gets out everytime the max-age expired, then it's served from cache for max-age seconds:

image

If you confirm that my understanding of hishel's behaviour is correct and most importantly that the obeserved behaviour is not correct, I can work on a a real patch (sync+async+matching tests).

Thanks a lot.

### PR DESCRIPTION
**Pull request by hartym**
Attempt to fix #239 (only async code, for now)

Before further work (aka cleanup, sync version, more tests maybe, maybe split tests ...), I'd like to validate a few things:

* I created a new test file. Should I keep it so or should I merge it in transport tests ?
* I mocked the clock using the test library "freezegun" which I'm used to and that did not require to understand the MockedClock behaviour. I don't know if having this (which is really handy ...) as a test dependencey would be ok or not
* Should I split the different test cases ?
* Do you prefer very explicit test code with duplicates, or should I factor the stored response extraction logic in a local function ?
* Any comments ?

### CODE REVIEW

**Comment by karpetrosyan** [comment]
> I created a new test file. Should I keep it so or should I merge it in transport tests ?

Let's add them into the `tests/_async/test_transport.py` and `tests/_async/test_pool.py`

> I mocked the clock using the test library "freezegun" which I'm used to and that did not require to understand the MockedClock behaviour. I don't know if having this (which is really handy ...) as a test dependencey would be ok or not

Actually, the idea behind using a clock was to avoid relying on third-party libraries solely for mocking the underlying time. Instead, we opted for dependency injection, which, in my opinion, is more explicit. So, let's use the clock for now.
The same applies to response mocking. Let's use the mechanism that is employed in other tests, utilizing the built-in `MockTransport`.

> Should I split the different test cases ?

I guess we only need a single test, for this case.

> Do you prefer very explicit test code with duplicates, or should I factor the stored response extraction logic in a local function ?

I prefer explicit test cases.

# Comments

We also support [HTTPCore](https://www.encode.io/httpcore/) connection pools as an alternative to high-level HTTPX transports. Therefore, the same changes should be made for `hishel/_async/_pool.py` and `tests/_async/test_pool.py`.

**Comment by hartym** [comment]
Noted, I'll apply the changes once I get a new time slot for this (probably this evening).

**Comment by karpetrosyan** [comment]
Also, you can refer to the [contribution documentation](https://hishel.com/contributing/). The maintenance process for hishel can be a bit confusing. We focus solely on writing asynchronous code, and the synchronous code is generated automatically. More information can be found in the documentation.

**Comment by hartym** [comment]
> Also, you can refer to the [contribution documentation](https://hishel.com/contributing/). The maintenance process for hishel can be a bit confusing. We focus solely on writing asynchronous code, and the synchronous code is generated automatically. More information can be found in the documentation.

Thanks. I did it first once complaining silently about why the heck the check script does not give me line numbers, but got it now.

I only have the pool version remaining and I think we'll be good to go. The autoincrementing Clock was a bit confusing as I do not really know how many times the clock is called by my test code, so I implemented one following the same logic but with a manual increment switch.

File: scripts/check [hunk 1]
```diff
@@ -6,7 +6,7 @@ if [ -d '.venv' ] ; then
 fi
 export SOURCE_FILES="hishel tests"
 
-${PREFIX}ruff format $SOURCE_FILES --diff
-${PREFIX}ruff check $SOURCE_FILES
+${PREFIX}ruff format $SOURCE_FILES
+${PREFIX}ruff check --fix $SOURCE_FILES
```
Comment (karpetrosyan) [inline scripts/check:10]: I think, you need to run ./scripts/lint for fixing

**Comment by karpetrosyan** [comment]
Oh, I just realized that we need to remove the cache entry if the 304 response is not cacheable. However, we currently do not support an API for removing entries from the storage. There's an ongoing [issue for introducing that API](https://github.com/karpetrosyan/hishel/issues/230)

I will work on it. Once done, we can add the removal part for cases where the 304 response contains something like no-store, and then get this merged.

File: scripts/check (hunk 1 above)
Comment (hartym) [reply scripts/check]: Yes, this was intended not to be commited, thanks for the lint-hint.

**Comment by hartym** [comment]
I think it's done (for this bug, not the additional issue you mentioned). Ready for final review :)

**Comment by hartym** [comment]
(of course, when it's ok, you can squash all the noise I generated)

**Comment by hartym** [comment]
Squashed my commits to remove my "out of conventions" commits from the history, merged upstream release.

**Comment by hartym** [comment]
> Do we really have unexpected behavior here? I'm not fully understanding what we are trying to fix.

My PR is ready, as far as I know.

I do have the unexpected behaviour described by the issue before my patch (and new tests agree with it).

I don't have it anymore after the patch (and new tests pass).

For me, the PR is ready for your review and merge, let me know if there is a misunderstanding somewhere.

**Comment by karpetrosyan** [comment]
> **Before patch:**
> 
> ```
> GET /cache/2 -> 200 with cache-control: public, max-age=2
> GET /cache/2 -> response from cache
> ... then 2 second later ...
> GET /cache/2 -> 200 with cache-control: public, max-age=2 (revalidation)
> GET /cache/2 -> 200 with cache-control: public, max-age=2 (revalidation)
> GET /cache/2 -> 200 with cache-control: public, max-age=2 (revalidation)
> (cache never used there anymore)
> ```
> 
> **After patch:**
> 
> ```
> GET /cache/2 -> 200 with cache-control: public, max-age=2
> GET /cache/2 -> response from cache
> ... then 2 second later ...
> GET /cache/2 -> 200 with cache-control: public, max-age=2 (revalidation)
> GET /cache/2 -> response from cache
> ```

Can we isolate this case? I have tried but seems, everything work fine.

**Comment by hartym** [comment]
It is isolated in the test cases added in the pull request. If you need, I can split the test but you said that's not necessary. The test fails before the patch.

https://github.com/karpetrosyan/hishel/blob/e447d22c9f91f96fe0ed189d1a2697dbc070627a/tests/_async/test_transport.py#L360

**Comment by karpetrosyan** [comment]
Let's look at https://github.com/karpetrosyan/hishel/issues/239#issuecomment-2179958037, seems like it should fail, but it works.

**Comment by hartym** [comment]
I missed this one, I need to got out now but will try that test as soon as I can.

**Comment by hartym** [comment]
Wrote isolated test case in https://github.com/karpetrosyan/hishel/issues/239#issuecomment-2188093351

**Comment by karpetrosyan** [comment]
> Wrote isolated test case in [#239 (comment)](https://github.com/karpetrosyan/hishel/issues/239#issuecomment-2188093351)

Let's fix test case in this PR, and I think we can get this merged.

**Comment by hartym** [comment]
Test fixed.

File: hishel/_sync/_pool.py [hunk 2]
```diff
@@ -134,6 +134,14 @@ def handle_request(self, request: Request) -> Response:
                 )
 
                 final_response.read()
+
+                if self._controller.is_cachable(request=request, response=final_response):
+                    metadata = Metadata(
```
Comment (karpetrosyan) [inline hishel/_sync/_pool.py:139]: I dont' think we need to create a new metadata here. The transport and pool implementations should be almost identical, in transports we do not create new metadata here

File: hishel/_async/_transports.py [hunk 3]
```diff
@@ -181,13 +181,24 @@ async def handle_async_request(self, request: Request) -> Response:
...
                 await final_httpcore_response.aread()
                 await revalidation_response.aclose()
 
                 assert isinstance(final_httpcore_response.stream, tp.AsyncIterable)
+
+                if self._controller.is_cachable(
+                    request=httpcore_request,
+                    response=final_httpcore_response,
+                ):
+                    await self._storage.store(
```
Comment (karpetrosyan) [inline hishel/_async/_transports.py:197]: I guess we need to check, if the response code is 200, then call `store` with the new metadata (by passing metadata=None so it creates automatically) but if the response code is 304 then we should use `update_metadata` method, which is for updating the entries, without touching the `ttl`. No we are reseting the metadata with each 304, which is not what we want

**Review by karpetrosyan — CHANGES_REQUESTED** [review]

File: tests/_async/test_pool.py [hunk 4]
```diff
@@ -254,3 +254,71 @@ async def test_pool_caching_post_method():
             response = await cache_pool.request("POST", "https://www.example.com", content=b"request-2")
             assert response.status == 200
             assert not response.extensions["from_cache"]
+
+
+@pytest.mark.anyio
+async def test_revalidation_with_new_content():
+    class MockedClock(BaseClock):
+        current = 1440504000  # Mon, 25 Aug 2015 12:00:01 GMT
```
Comment (karpetrosyan) [inline tests/_async/test_pool.py:262]:
```suggestion
        current = 1440504000  # Mon, 25 Aug 2015 12:00:00 GMT
```

File: hishel/_sync/_pool.py (hunk 2 above)
Comment (hartym) [reply hishel/_sync/_pool.py:139]: I did a simple copy of the storage logic few lines below, but you're absolutely right. I'm just wondering if we should (in both cases transport or pool) increment number of uses or not there, or amend the metadata in any way before re-store.

Comment (karpetrosyan) [reply hishel/_sync/_pool.py:139]: I guess it's already handled by create_hishel_response method

File: hishel/_async/_transports.py (hunk 3 above)
Comment (hartym) [reply hishel/_async/_transports.py:197]: Ok, I see that _create_hishel_response already does the "update_metadata" call so I simply exclude 304 from the condition to store revalidation response, as it is indeed already done in this case. But maybe if the _create_hishel_response already have the responsibility of updating stored entry, the storage.store calls would benefit from being in the same place ?

File: hishel/_sync/_pool.py (hunk 2 above)
Comment (hartym) [reply hishel/_sync/_pool.py:139]: I mean in the case of not a 304 (cached==False)

**Comment by hartym** [comment]
Fixes based on your comments, I removed the double metadata creation in pool and transport that was hanging around (the store method does it, already). I hope I did not miss something there, let me know.

**Comment by karpetrosyan** [comment]
Thanks for the PR :pray:  It's very neat!

**Review by karpetrosyan — APPROVED** [review]

**Pull request merged**